### ISSUE TITLE
Sanitizing parsers

### ISSUE THREAD

**Issue by peterbrittain** [peterbrittain/asciimatics#224]
**Is your feature request related to a problem? Please describe.**
Several people have seen unexpected behaviour due to control codes that have been embedded in the data that they set in a widget - e.g. putting `\r` characters in the data.  This causes all sorts of odd layout problems displaying the widget (as the control codes change location on the screen to something unexpected).

**Describe the solution you'd like**
Now that we have parsers, how about creating one that simply replaces all control codes with a readable version - e.g. "^M" for `\r`, etc.  Simply add this parser to the widget when you construct it and you never see the issue again!

Added bonus is that this will still maintain the control codes in the data stream should you need to recreate them afterwards.

**Comment by peterbrittain** [comment]
This is actually pretty simple.  You just need to take the logic from an existing parser and convert any control codes - ordinal values 0-31 - to equivalent escaped text strings.  

Possibly a good thing to add to the AnsiTerminalParser - as I expect a lot of the junk to come from terminal output that is copied to the widget one way or another.

**Comment by edoffagne** [comment]
Hello,
I would be happy to give it a try.
How would you see it ? Something like:
`parser = sanitizeParser()`
`tokens = parser.parse("\r")`
where "\r" would become "^M" ?

**Comment by peterbrittain** [comment]
Great!  Yes - that's the idea...  

You'll see that in the `advanced_terminal` branch I've already created a more advanced AnsiTerminalParser that happens to handle this particular control code, so I'm inclined to make this a separate parser now (rather than extend the existing parser).  The mainline logic and test cases should be pretty easy to extract from the existing parsers.

**Comment by edoffagne** [comment]
OK thanks.
I will have a look at the branch `advanced_terminal`

**Comment by peterbrittain** [comment]
That branch is still under development and so please, by all means, have a look and play, but don't rely on the API staying the same by the time it is merged into master.

**Comment by edoffagne** [comment]
@peterbrittain I have written a skeleton of the parser and associated tests that for now just checks for carriage return. Before to go further I wanted to check that it makes sense.
https://github.com/edoffagne/asciimatics/tree/parser-sanitize

**Comment by peterbrittain** [comment]
You're getting there.  The idea of the parser is that it is a generator function that consumes the supplied text and yields the transformed text.

That means that you want to look at each character in the supplied `text`, decide if you can just yield that character, or yield something else instead of it.  So the bsic logic for a sanitizing parser is:
```
for each letter in text
    if letter is a control code (ordinal value less than 32):
        yield "^" + chr(ord("@") + ord(letter))
    else:
        yield letter
````

I hope that makes senss.

**Comment by edoffagne** [comment]
Thanks for the feedback.
Yes it makes sense. I will update the code in that direction and submit a PR once I have something ready.

**Comment by peterbrittain** [comment]
Implemented by #229
